src/main/java/agh.ics.oop_project1/SimulationEngine.java

```diff
@@ -0,0 +1,253 @@
+package agh.ics.oop_project1;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class SimulationEngine implements Runnable, IDayChangeObserver {
+
+    private IWorldMap map;
+    private int numberOfAnimals;
+    private final ArrayList<Animal> animals = new ArrayList<>();
+    private HashMap<String,Integer> mostPopularGenotypes = new HashMap<>();
+    private int lengthOfGenome;
+    private int animalCounter = 0;
+    private final int floraGrowth;
+    private int nutritionValue;
+    private int currentDay = 0;
+    private int energyNeededToBreed;
+    private int energyLossOnBreeding;
+    private int minMutations;
+    private int maxMutations;
+    private long daysLivedByDead = 0;
+    private long deadCount = 0;
+    private final ArrayList<IDayChangeObserver> dayChangeObservers = new ArrayList<>();
+    private boolean isWaiting = false;
+
+    public SimulationEngine(ConfigurationManager configmanager) {
+        if(configmanager.getMapVariant().equals("globe")) {
+            this.map = new GlobeMap(configmanager.getMapHeight(), configmanager.getMapWidth());
+        }
+        else if(configmanager.getMapVariant().equals("nether")) {
+            this.map = new NetherMap(configmanager.getMapHeight(), configmanager.getMapWidth(), configmanager.getEnergyDrainAmount());
+        }
+
+        this.numberOfAnimals = configmanager.getNumberOfAnimals();
+        this.floraGrowth = configmanager.getFloraGrowth();
+        this.nutritionValue = configmanager.getNutritionValue();
+        this.energyNeededToBreed = configmanager.getEnergyNeededToBreed();
+        this.energyLossOnBreeding = configmanager.getEnergyLossOnBreeding();
+        this.minMutations = configmanager.getMinMutations();
+        this.maxMutations = configmanager.getMaxMutations();
+        this.lengthOfGenome = configmanager.getLengthOfGenome();
+        for(int i = 0; i < numberOfAnimals; i++) {
+            Animal animal = new Animal(map, configmanager.getLengthOfGenome(), configmanager.getInitialEnergy(), animalCounter, currentDay);
+            animals.add(animal);
+            animal.addObserver(this.map);
+            map.placeAnimal(animal);
+            animalCounter++;
+            String genomeString = animal.getGenome().stream().map(Object::toString)
+                    .collect(Collectors.joining());
+            if(!mostPopularGenotypes.containsKey(genomeString)) {
+                mostPopularGenotypes.put(genomeString,1);
+            }
+            else {
+                mostPopularGenotypes.put(genomeString, mostPopularGenotypes.remove(genomeString)+1);
+            }
+        }
+
+        this.map.putFlora(configmanager.getStartingFlora());
+    }
+
+    public SimulationEngine(IWorldMap map, int numberOfAnimals, int initialEnergy, int startingFlora, int floraGrowth, int nutritionValue, int energyNeededToBreed, int energyLossOnBreeding, int minMutations, int maxMutations, int lengthOfGenome) {
```

Comment:
tu mamy znów bardzo odseparowane konstruktory - przykład dlaczego przydałoby się mieć klasę przechowującą konfigurację, wtedy tak naprawdę `SimulationEngine` miałoby jeden konstruktor. A tak to kod robi się tu bardzo poplątany, powtarzają się jego fragmenty, jest to brzydkie i niebezpieczne. Szczerze mówiąc, nie wiem czy nawet nie ma tu gdzieś buga, bo jak próbuję wczytać symulację z pliku to po uruchomieniu nie widzę zwierząt na mapie.